Add `cilium bpf template` CLI

joestringer opened the pull request:
Based on ~~#7095~~, ~~#7347~~.

Add some visibility tooling around the recent BPF templating work.

* Add links from `/var/run/cilium/state/EPID/template.o` to `/var/run/cilium/state/templates/xxx/template.o` so that the link between SHA and endpoint can be observed.

* Add `cilium bpf sha list` (alias `cilium bpf template list`)

   ```
   $ daemon/cilium bpf sha list
   SHA                                        Endpoint(s)
   <No template used>                         142
   <Template path invalid>                    1747
                                              3147
   42041af982a4a878c5efbfae04dc23fc68d4deb5   4076
   ```

   Once #7095 is merged, all entries in this output should consist of a SHA followed by a list of endpoints; 
   I included the above output to show what happens when a template is not used (a proxy for whether 
   there is a template.o file in the `/var/run/cilium/state/EPID` dir), or when the template symlink is broken.

 * Add `cilium bpf sha describe <sha>`

   This is just a convenience method for printing the template LXC file.

   ```
   $ daemon/cilium bpf sha get 42041af982a4a878c5efbfae04dc23fc68d4deb5
   #include "lib/utils.h"

   DEFINE_IPV6(LXC_IP, 0x20, 0x1, 0xdb, 0x8, 0xb, 0xad, 0xca, 0xfe, 0x60, 0xd, 0xbe, 0xe2, 0xb, 0xad, 0xca, 0xfe);                                                                                               
   DEFINE_U32(LXC_IPV4, 0x030200c0);       /* 50462912 */
   #define LXC_IPV4 fetch_u32(LXC_IPV4)
   ...
   ```

* Each time a template is generated, add an fsnotify hook on the BPF template object and if it's removed, remove it from the cache inside Cilium. Subsequent regenerations will recreate the template.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7297)
<!-- Reviewable:end -->

joestringer committed:
loader: Symlink the template path in each statedir

Symlink the template used in each endpoint's state directory to improve
debuggability.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
cilium: Add bpf template CLI

* cilium bpf sha list

  $ daemon/cilium bpf sha list
  SHA                                        Endpoint(s)
  <No template used>                         142
  <Template path invalid>                    1747
                                             3147
  42041af982a4a878c5efbfae04dc23fc68d4deb5   4076

* cilium bpf sha describe <sha>

  $ daemon/cilium bpf sha get 42041af982a4a878c5efbfae04dc23fc68d4deb5
  #include "lib/utils.h"

  DEFINE_IPV6(LXC_IP, 0x20, 0x1, 0xdb, 0x8, 0xb, 0xad, 0xca, 0xfe, 0x60, 0xd, 0xbe, 0xe2, 0xb, 0xad, 0xca, 0xfe);
  DEFINE_U32(LXC_IPV4, 0x030200c0);       /* 50462912 */
  #define LXC_IPV4 fetch_u32(LXC_IPV4)
  ...

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Watch filesystem for template changes

This allows us to support deleting the template objects from the
filesystem, then next time the daemon attempts to use the template it
will recompile the template first, then cache the newly compiled
template.

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon commented on cilium/cmd/bpf_sha_list.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+var (
+	stateDir     = filepath.Join(defaults.RuntimePath, defaults.StateDir)
```
`RunDir` may not always be this value. maybe provide an option to override it if necessary in case the user has overriden it in the daemon's configuration? This use case isn't common, but it is possible.

joestringer replied:
This isn't the only tool with this problem, we'd need to fix it up for multiple cli commands. That said it's such an unusual case that I don't currently intend on working on this.

ianvernon reviewed (approved):
LGTM - the comment I made above can be a change in a follow-up ticket.

joestringer commented on pkg/datapath/loader/cache.go:
```diff
@@ -111,17 +116,24 @@ func newObjectCache(dp datapath.Datapath, nodeCfg *datapath.LocalNodeConfigurati
 	oc := &objectCache{
 		Datapath:         dp,
 		workingDirectory: workingDir,
+		newTemplates:     make(chan string),
```
This likely needs to be buffered. /cc @tgraf

joestringer replied:
(or alternatively we need some kind of timeout on the send side to figure out when shutdown is occurring), see the `default:` case where this channel is populated in `insert()`.

joestringer replied:
Top patch resolves this concern.

joestringer committed:
loader: Queue template filepaths to watch

Queue the template paths to watch and signal when the template watcher
dies. It's still plausible that the template is created and deleted
before the templateWatcher starts watching it, but the possibility is
low and there's probably not a lot we can do at that point.

Signed-off-by: Joe Stringer <joe@cilium.io>
